src/NHibernate/Cache/Entry/StructuredCollectionCacheEntry.cs

```diff
@@ -11,7 +11,7 @@ public class StructuredCollectionCacheEntry : ICacheEntryStructure
 		public virtual object Structure(object item)
 		{
 			var entry = (CollectionCacheEntry)item;
-			return new List<object>(entry.State);
+			return new List<object>((object[])entry.DisassembledState);
```

Comment:
So, the `//TODO: assumes all collections disassemble to an array!` has moved here indeed.